### ISSUE TITLE
Failed to build gem native extension

### ISSUE THREAD

**Issue by BR0kEN-** [geerlingguy/ansible-role-ruby#18]
Hi, Jeff.

I found an error on CentOS 6.6 and Ruby role. Here is the error log:

```
TASK: [geerlingguy.ruby | Install configured gems.] ***************************
failed: [185.86.76.75] => (item=compass) => {"cmd": "/usr/bin/gem install --user-install --no-rdoc --no-ri compass", "failed": true, "item": "compass", "rc": 1}
stderr: WARNING:  You don't have /root/.gem/ruby/1.8/bin in your PATH,
      gem executables will not run.
ERROR:  Error installing compass:
    ERROR: Failed to build gem native extension.

    /usr/bin/ruby -r ./siteconf20150624-25661-xtsd0c-0.rb extconf.rb
mkmf.rb can't find header files for ruby at /usr/lib/ruby/ruby.h

extconf failed, exit code 1

Gem files will remain installed in /root/.gem/ruby/1.8/gems/ffi-1.9.8 for inspection.
Results logged to /root/.gem/ruby/1.8/extensions/x86_64-linux/1.8/ffi-1.9.8/gem_make.out

stdout: Building native extensions.  This could take a while...

msg: WARNING:  You don't have /root/.gem/ruby/1.8/bin in your PATH,
      gem executables will not run.
ERROR:  Error installing compass:
    ERROR: Failed to build gem native extension.

    /usr/bin/ruby -r ./siteconf20150624-25661-xtsd0c-0.rb extconf.rb
mkmf.rb can't find header files for ruby at /usr/lib/ruby/ruby.h

extconf failed, exit code 1

Gem files will remain installed in /root/.gem/ruby/1.8/gems/ffi-1.9.8 for inspection.
Results logged to /root/.gem/ruby/1.8/extensions/x86_64-linux/1.8/ffi-1.9.8/gem_make.out

FATAL: all hosts have already failed -- aborting
```

My configuration is the following:

``` yaml
# Ruby.
ruby_install_gems:
  - compass
```

**Comment by jorgegc** [comment]
Hi, I am having a similar issue on Ubuntu 14.04 with Ruby enabled. Here is the error log:

```
Gem::Installer::ExtensionBuildError: ERROR: Failed to build gem native extension.

        /usr/bin/ruby1.9.1 extconf.rb
checking for clock_gettime() in -lrt... yes
creating Makefile

make
compiling hitimes.c
compiling hitimes_instant_clock_gettime.c
compiling hitimes_instant_osx.c
compiling hitimes_instant_windows.c
compiling hitimes_interval.c
compiling hitimes_stats.c
linking shared-object hitimes/1.9/hitimes.so

make install
/usr/bin/install -c -m 0755 hitimes.so /var/www/vendor/ruby/1.9.1/gems/hitimes-1.2.2/lib/hitimes/1.9
/usr/bin/install: setting permissions for ‘/var/www/vendor/ruby/1.9.1/gems/hitimes-1.2.2/lib/hitimes/1.9/hitimes.so’: Operation not permitted
make: *** [/var/www/vendor/ruby/1.9.1/gems/hitimes-1.2.2/lib/hitimes/1.9/hitimes.so] Error 1
```

I find it strange that I am getting this "Operation not permitted" when clearly I have permission to modify the contents of that folder.

Any help would be greatly appreciated. Thanks!

**Comment by richlewis14** [comment]
@jorgegc did you ever resolve your issue? I am getting the same error

**Comment by jorgegc** [comment]
No, I still haven't resolved the issue @richlewis14. I switched to installing the Gems globally as a workaround it but I would love to get to the bottom of it.

**Comment by geerlingguy** [comment]
Can't promise any time to look into this in the short term, but I'll try to get to it... right now I'm focused on LAMP/LEMP stack-related roles, and will get back to triaging/fixing Ruby and some of the other language roles hopefully by end-of-year. If anyone else has any ideas or code to help fix this, I'm all ears :)

**Comment by geerlingguy** [comment]
Still didn't get much time to devote to this, but in some 'happy path' testing on CentOS 7 and Ubuntu 14.04, it seems things are still working okay for me :/

**Comment by geerlingguy** [comment]
Not sure about CentOS 6 (since I'm not officially testing on that version now), but for the others, this should be resolved via the work done in https://github.com/geerlingguy/ansible-role-ruby/issues/38

Basically, many gems required `make` and other build tools to be present. And even old faithful standbys like `sass` are now built instead of just downloaded/installed... therefore I figured it's worth including `build-essential` (Debian) or `@development` (RedHat) by default.

So, closing this issue as I think it should be resolved with this role, version 2.4.3+.
